File: Delphi/ComponentEditors.pas
unit ComponentEditors;

interface

uses
  System.Classes, System.TypInfo,
  DesignIntf, DesignEditors;

type
  TPropEditorWithDialog = class
  private
    FPropName: string;
    procedure CheckEditProperty(const Prop: IProperty);
  public
    procedure EditProperty(Component: TPersistent; const PropName: string;
      const Designer: IDesigner);
  end;

procedure EditPropertyDlg(Component: TPersistent; const PropName: string;
  const Designer: IDesigner);

implementation

type
  TDesignerSelectionsAccess = class(TDesignerSelections);

{ TPropEditorWithDialog }

procedure TPropEditorWithDialog.CheckEditProperty(const Prop: IProperty);
begin
  if Prop.GetName = FPropName then
    Prop.Edit;
end;

procedure TPropEditorWithDialog.EditProperty(Component: TPersistent;
  const PropName: string; const Designer: IDesigner);
var
  Selections: TDesignerSelections;
  ISelections: IDesignerSelections;
begin
  FPropName := PropName;
  Selections := TDesignerSelections.Create;
  ISelections := Selections;
  TDesignerSelectionsAccess(Selections).Add(Component);
  GetComponentProperties(ISelections, [tkClass], Designer, CheckEditProperty);
end;

{ EditPropertyDlg }

procedure EditPropertyDlg(Component: TPersistent; const PropName: string;
  const Designer: IDesigner);
var
  Edt: TPropEditorWithDialog;
begin
  Edt := TPropEditorWithDialog.Create;
  try
    Edt.EditProperty(Component, PropName, Designer);
  finally
    Edt.Free;
  end;
end;

end.

File: Cpp/ComponentEditors.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include <memory>

#include "ComponentEditors.h"

//#include <boost/shared_ptr.hpp>

using std::unique_ptr;

//---------------------------------------------------------------------------

#pragma package(smart_init)

//---------------------------------------------------------------------------
namespace Componenteditors {
//---------------------------------------------------------------------------

void __fastcall TPropEditorWithDialog::CheckEditProperty( _di_IProperty const Prop )
{
    if ( Prop->GetName() == propName_ )
        Prop->Edit();
}
//---------------------------------------------------------------------------

void __fastcall TPropEditorWithDialog::EditProperty( TPersistent* Component,
                                                     String const PropName,
                                                     _di_IDesigner const Designer )
{
    TDesignerSelectionsPub* Components = new TDesignerSelectionsPub();
    propName_ = PropName;
    Components->Add( Component );
    GetComponentProperties(
        Components->operator _di_IDesignerSelections(),
        Typinfo::TTypeKinds() << tkClass,
        Designer,
        &CheckEditProperty
    );
}

void EditPropertyDlg( TPersistent* Component, String PropName,
                      _di_IDesigner Designer )
{
/*
    boost::shared_ptr<TPropEditorWithDialog>
    Edt( new TPropEditorWithDialog );

    Edt->EditProperty( Component, PropName, Designer );
*/
    unique_ptr<TPropEditorWithDialog> Edt( new TPropEditorWithDialog );
    Edt->EditProperty( Component, PropName, Designer );
}

//---------------------------------------------------------------------------
} // end of namespace Componenteditors
//---------------------------------------------------------------------------


